Packages/ChatSecure/ChatSecure/Classes/Networking/Services/Group/GroupService.swift

```diff
@@ -0,0 +1,230 @@
+			guard let server = server, let profile = server.profile else {
+				return .success([])
+			}
+			let groups = await convertGroupFromResponse(groups: data, server: server)
+			await channelStorage.realmManager.addAndUpdateGroups(groups: groups)
+			return .success(groups)
+
+		case .failure(let error):
+			return .failure(error)
+		}
+	}
+	
+	public func joinGroup(by groupId: Int64, domain: String) async -> (Result<Group_BaseResponse, Error>) {
+		let apiService = channelStorage.getChannel(domain: domain)
+		guard let clientId = apiService.owner?.id else { return .failure(ServerError.unknown) }
+		var request = Group_JoinGroupRequest()
+		request.groupID = groupId
+		request.clientID = clientId
+		
+		return await apiService.joinGroup(request)
+	}
+	
+	public func addMember(_ user: Group_ClientInGroupObject, groupId: Int64, domain: String) async -> (Result<Group_BaseResponse, Error>) {
+		let apiService = channelStorage.getChannel(domain: domain)
+		guard let clientId = apiService.owner?.id,
+			  let userName = apiService.owner?.displayName else { return .failure(ServerError.unknown) }
+		var requestAddedMember = Group_MemberInfo()
+		requestAddedMember.id = user.id
+		requestAddedMember.workspaceDomain = user.workspaceDomain
+		requestAddedMember.displayName = user.displayName
+		requestAddedMember.status = ""
+		
+		var requestAddingMember = Group_MemberInfo()
+		requestAddingMember.id = clientId
+		requestAddingMember.workspaceDomain = domain
+		requestAddingMember.displayName = userName
+		requestAddingMember.status = ""
+		requestAddingMember.refGroupID = groupId
+		
+		var request = Group_AddMemberRequest()
+		request.addingMemberInfo = requestAddingMember
+		request.addedMemberInfo = requestAddedMember
+		request.groupID = groupId
+		
+		return await apiService.addMember(request)
+	}
+	
+	public func leaveGroup(_ user: Group_ClientInGroupObject, groupId: Int64, domain: String) async -> (Result<Group_BaseResponse, Error>) {
+		let apiService = channelStorage.getChannel(domain: domain)
+		guard let clientId = apiService.owner?.id,
+			  let userName = apiService.owner?.displayName else { return .failure(ServerError.unknown) }
+		var memberInfo = Group_MemberInfo()
+		memberInfo.id = user.id
+		memberInfo.workspaceDomain = user.workspaceDomain
+		memberInfo.displayName = user.displayName
+		memberInfo.status = ""
+		memberInfo.refGroupID = groupId
+
+		var memberBy = Group_MemberInfo()
+		memberBy.id = clientId
+		memberBy.workspaceDomain = domain
+		memberBy.displayName = userName
+		memberBy.status = ""
+		memberBy.refGroupID = groupId
+
+		var request = Group_LeaveGroupRequest()
+		request.leaveMember = memberInfo
+		request.leaveMemberBy = memberInfo
+		request.groupID = groupId
+		
+		return await channelStorage.getChannel(domain: domain).leaveGroup(request)
+	}
+}
+
+private extension GroupService {
+	func convertGroupFromResponse(groups: Group_GetJoinedGroupsResponse, server: RealmServer) async -> [RealmGroup] {
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ClearKeep/ck-ios/blob/8649f3f2f0ce9de1db3f8c96a6cf34266ff6e2d9/Packages/ChatSecure/ChatSecure/Classes/Networking/Services/Group/GroupService.swift#L153"></span>Function body should span 50 lines or less excluding comments and whitespace: currently spans 56 lines<br />
<code>function_body_length</code> <code>GroupService.swift:153</code></td>
    </tr>
  </tbody>
</table>